homeassistant/core.py

```diff
@@ -1374,30 +1394,56 @@ def async_fire(
             raise MaxLengthExceeded(
                 event_type, "event_type", MAX_LENGTH_EVENT_EVENT_TYPE
             )
+        return self._async_fire(event_type, event_data, origin, context, time_fired)
 
-        listeners = self._listeners.get(event_type, [])
-        match_all_listeners = self._match_all_listeners
+    @callback
+    def _async_fire(
```

Comment:
With latest version its only 1.15x over baseline which is really good considering we are calling this almost 10x as much